Kernel/Thread.h

```diff
@@ -824,7 +824,12 @@ class Thread
             return EDEADLK;
 
         SpinlockLocker lock(m_lock);
-        if (!m_is_joinable || state() == Thread::State::Dead)
+
+        // Joining dead threads is allowed for two main reasons:
+        // - Thread join behavior should not be racy when a thread is joined and exiting at roughly the same time.
+        //   This is common behavior when threads are given a signal to end (meaning they are going to exit ASAP) and then joined.
+        // - POSIX requires that exited threads are joinable (at least, there is no language in the specification forbidding it).
+        if (!m_is_joinable || state() == Thread::State::Invalid)
```

Comment:
This is a bit different from the recommendations in the Rationale section of Dr. Posix:

> If an implementation detects that the value specified by the thread argument to pthread_join() does not refer to a joinable thread, it is recommended that the function should fail and report an [EINVAL] error.
>
> If an implementation detects that the value specified by the thread argument to pthread_join() refers to the calling thread, it is recommended that the function should fail and report an [EDEADLK] error.
> 
> If an implementation detects use of a thread ID after the end of its lifetime, it is recommended that the function should fail and report an [ESRCH] error.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html

It sounds like it's recommending that we check the thread state here: if it's dead, we should return ESRCH. if it's not a joinable thread, we should return EINVAL.

Reply:
If a thread is dead or dying, we shouldn't block the calling thread until it gets reaped by the Finalizer thread. Not sure if this implementation will do that, but it seems like the posix recommendation makes sense to me.

Reply:
Hmmm interesting, I either had an old version of the POSIX standard open or missed that last paragraph; I was confident I scanned this page several times for anything relating to this. I'm still asking myself how that makes any sense; if I'm joining a finite-lifespan thread, it always (given enough "allowed" scheduler unfairness) has the possibility of exiting before I can call join(). We should get ESRCH left and right with normal "do one task then exit" threads, right?

Reply:
Either way, I think will add another commit that ignores ESRCH in our Core::Thread which doesn't have to adhere to POSIX semantics.

Reply:
See Discord; after re-checking I think the new behavior is correct.